Create and restore plan snapshots

duranb opened the pull request:
resolves #876 

To test:
1. Create a plan with some activities
2. Click on the plan name on the nav bar to open the menu
3. Click on "Take Snapshot"
4. Fill in the snapshot name and click "Create Snapshot" button
5. Add more activities to the plan
6. Add the `snapshotId` search parameter with the snapshot ID number (e.g. `snapshotId=1`) to the url and hit enter
7. Verify that the correct activities are loaded associated with the snapshot (should be fewer activities)

duranb committed:
add snapshot creation modal
add snapshot restore modal
load snapshot activities if snapshotId search parameter is provided

duranb committed:
add search parameter enum

duranb committed:
add test

AaronPlave commented on .eslintrc.cjs:
```diff
@@ -36,6 +36,7 @@ module.exports = {
     ],
     '@typescript-eslint/no-explicit-any': 'off',
     '@typescript-eslint/no-inferrable-types': 'off',
+    '@typescript-eslint/no-unused-vars': ['error', { ignoreRestSiblings: true, argsIgnorePattern: '^_' }],
```
What is this one for?

duranb replied:
This basically let's us write code like

``` const { foo: _ignore, ...restOfStuff } = bar```

and eslint won't complain about `_ignore` not being used if we intend to delete it from the object.

AaronPlave replied:
Ah i see, cool.

AaronPlave commented on src/components/menus/PlanMenu.svelte:
```diff
@@ -76,6 +84,13 @@
       {/if}
+      <hr class="menu-divider" />
+      <MenuItem on:click={createPlanSnapshot}>
+        <div class="column-name">Take Snapshot</div>
```
Create or Take? Seeing Create in the modal, perhaps they should be unified?

duranb replied:
Ah yeah. I was just going by the figma wording. @lklyne do you have a preference?

AaronPlave commented on src/components/modals/CreatePlanSnapshotModal.svelte:
Any sense in describing what a snapshot is here or is that weird?
<img width="335" alt="image" src="https://github.com/NASA-AMMOS/aerie-ui/assets/4419607/9874677d-d727-4dba-8e1f-f4bd52562be8">

AaronPlave replied:
@lklyne

lklyne replied:
Yes -- input from users and design direction suggests we should have a "comments" section here. Not sure if that's supported on the backend yet, but it's the goal.

AaronPlave replied:
Oh I was thinking more like what is a snapshot? Like what is going to happen when you click on "create snapshot"

AaronPlave commented on src/utilities/modal.ts:
```diff
@@ -441,6 +478,47 @@ export async function showPlanMergeRequestsModal(
+/**
+ * Shows a RestorePlanSnapshotModal with the supplied arguments.
+ */
+export async function showRestorePlanSnapshotModal(
```
Is this for the next PR? Cool to keep it in but just wanted to make sure it wasn't currently accessible?

duranb replied:
Yep! I tried to hopefully set you up as much as possible for the panel work.

AaronPlave replied:
:keanu-thanks:

AaronPlave commented on src/routes/plans/[id]/+page.svelte:
```diff
@@ -137,6 +141,16 @@
       .getResourceTypes($plan.model_id, data.user)
       .then(initialResourceTypes => ($resourceTypes = initialResourceTypes));
   }
+  $: if (data.initialPlanSnapshotId !== null) {
```
Do we care if the snapshotId doesn't map to a snapshot? Like should we warn the user if the snapshot does not exist? Otherwise you might think you're viewing the snapshot. Though I guess the design call for that horizontal message conveying that you're viewing a snapshot so maybe that's enough?

duranb replied:
hmm yeah! that's a good idea. I still have more work to do for snapshots within the plan page so I could add that message in my next PR?

AaronPlave replied:
Sounds good to me

AaronPlave reviewed (approved):
LGTM, just a few comments/questions.

duranb committed:
update create modal title to match button
add description of snapshot to creation modal
prepopulate snapshot name with plan name
